File: packages/sdk-java/src/main/java/org/openinherit/sdk/Validator.java
package org.openinherit.sdk;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.networknt.schema.JsonSchema;
import com.networknt.schema.JsonSchemaFactory;
import com.networknt.schema.SpecVersion;
import com.networknt.schema.ValidationMessage;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class Validator {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final JsonSchema ESTATE_SCHEMA;
    private static final JsonSchema CATALOGUE_SCHEMA;

    // Catalogue $schema URI used for auto-detection
    private static final String CATALOGUE_SCHEMA_URI = "https://openinherit.org/v3/catalogue.json";

    static {
        try {
            JsonSchemaFactory factory = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V202012);

            try (InputStream estateStream = Validator.class.getResourceAsStream("/inherit-v3-bundled.json")) {
                if (estateStream == null) {
                    throw new IllegalStateException("inherit-v3-bundled.json not found on classpath");
                }
                JsonNode estateNode = MAPPER.readTree(estateStream);
                ESTATE_SCHEMA = factory.getSchema(estateNode);
            }

            try (InputStream catalogueStream = Validator.class.getResourceAsStream("/catalogue-v3-bundled.json")) {
                if (catalogueStream == null) {
                    throw new IllegalStateException("catalogue-v3-bundled.json not found on classpath");
                }
                JsonNode catalogueNode = MAPPER.readTree(catalogueStream);
                CATALOGUE_SCHEMA = factory.getSchema(catalogueNode);
            }
        } catch (Exception e) {
            throw new ExceptionInInitializerError(e);
        }
    }

    /**
     * Validates an INHERIT document, auto-detecting whether it is an estate or catalogue
     * document by inspecting the {@code $schema} field. Defaults to estate mode if the
     * field is absent or does not match the catalogue URI.
     *
     * <p>This performs Level 1 (JSON Schema) validation only. The returned
     * {@link ValidationResult#conformanceLevel()} will be 0 if schema validation
     * fails, or 1 if it passes.
     *
     * @param jsonDocument the INHERIT document as a JSON string
     * @return a {@link ValidationResult} describing whether the document is valid
     */
    public static ValidationResult validate(String jsonDocument) {
        try {
            JsonNode doc = MAPPER.readTree(jsonDocument);
            String schemaUri = doc.path("$schema").asText("");
            if (CATALOGUE_SCHEMA_URI.equals(schemaUri)) {
                return runValidation(doc, CATALOGUE_SCHEMA, "catalogue");
            } else {
                return runValidation(doc, ESTATE_SCHEMA, "estate");
            }
        } catch (Exception e) {
            List<ValidationResult.ValidationError> errors = List.of(
                new ValidationResult.ValidationError("$", "Failed to parse JSON: " + e.getMessage())
            );
            return new ValidationResult(false, "estate", errors, ValidationResult.DISCLAIMER, 0);
        }
    }

    /**
     * Validates an INHERIT document against the catalogue schema, regardless of the
     * {@code $schema} field present in the document.
     *
     * @param jsonDocument the INHERIT catalogue document as a JSON string
     * @return a {@link ValidationResult} describing whether the document is valid
     */
    public static ValidationResult validateCatalogue(String jsonDocument) {
        try {
            JsonNode doc = MAPPER.readTree(jsonDocument);
            return runValidation(doc, CATALOGUE_SCHEMA, "catalogue");
        } catch (Exception e) {
            List<ValidationResult.ValidationError> errors = List.of(
                new ValidationResult.ValidationError("$", "Failed to parse JSON: " + e.getMessage())
            );
            return new ValidationResult(false, "catalogue", errors, ValidationResult.DISCLAIMER, 0);
        }
    }

    /**
     * Validates an INHERIT document against the estate schema, regardless of the
     * {@code $schema} field present in the document.
     *
     * @param jsonDocument the INHERIT estate document as a JSON string
     * @return a {@link ValidationResult} describing whether the document is valid
     */
    public static ValidationResult validateEstate(String jsonDocument) {
        try {
            JsonNode doc = MAPPER.readTree(jsonDocument);
            return runValidation(doc, ESTATE_SCHEMA, "estate");
        } catch (Exception e) {
            List<ValidationResult.ValidationError> errors = List.of(
                new ValidationResult.ValidationError("$", "Failed to parse JSON: " + e.getMessage())
            );
            return new ValidationResult(false, "estate", errors, ValidationResult.DISCLAIMER, 0);
        }
    }

    /**
     * Performs Level 2 validation on an INHERIT estate document.
     *
     * <p>Level 2 first runs Level 1 (JSON Schema) validation. If that passes,
     * it additionally checks that all UUID cross-references between entities
     * resolve to an actual entity in the document. For example, if
     * {@code estate.testatorPersonId} is "abc-123", there must be a person
     * with {@code id: "abc-123"} in the {@code people} array.
     *
     * <p>The returned {@link ValidationResult#conformanceLevel()} will be:
     * <ul>
     *   <li>0 — Level 1 (schema) validation failed</li>
     *   <li>1 — Level 1 passed but Level 2 (referential integrity) failed</li>
     *   <li>2 — Both levels passed</li>
     * </ul>
     *
     * @param jsonDocument the INHERIT estate document as a JSON string
     * @return a {@link ValidationResult} describing the outcome
     */
    public static ValidationResult validateLevel2(String jsonDocument) {
        JsonNode doc;
        try {
            doc = MAPPER.readTree(jsonDocument);
        } catch (Exception e) {
            List<ValidationResult.ValidationError> errors = List.of(
                new ValidationResult.ValidationError("$", "Failed to parse JSON: " + e.getMessage())
            );
            return new ValidationResult(false, "estate", errors, ValidationResult.DISCLAIMER, 0);
        }

        // Determine schema mode (only estate supports Level 2)
        String schemaUri = doc.path("$schema").asText("");
        String mode = CATALOGUE_SCHEMA_URI.equals(schemaUri) ? "catalogue" : "estate";
        JsonSchema schema = CATALOGUE_SCHEMA_URI.equals(schemaUri) ? CATALOGUE_SCHEMA : ESTATE_SCHEMA;

        // Level 1: schema validation
        ValidationResult level1 = runValidation(doc, schema, mode);
        if (!level1.valid()) {
            // conformanceLevel is already 0 from runValidation when there are errors
            return level1;
        }

        // Level 2: referential integrity
        List<ValidationResult.ValidationError> refErrors = ReferenceChecker.check(doc);
        if (!refErrors.isEmpty()) {
            return new ValidationResult(false, mode, refErrors, ValidationResult.DISCLAIMER, 1);
        }

        return new ValidationResult(true, mode, List.of(), ValidationResult.DISCLAIMER, 2);
    }

    private static ValidationResult runValidation(JsonNode doc, JsonSchema schema, String mode) {
        Set<ValidationMessage> messages = schema.validate(doc);
        List<ValidationResult.ValidationError> errors = new ArrayList<>();
        for (ValidationMessage msg : messages) {
            String path = msg.getInstanceLocation().toString();
            String message = msg.getMessage();
            errors.add(new ValidationResult.ValidationError(path, message));
        }
        boolean valid = errors.isEmpty();
        int conformanceLevel = valid ? 1 : 0;
        return new ValidationResult(valid, mode, errors, ValidationResult.DISCLAIMER, conformanceLevel);
    }
}

File: packages/sdk-csharp/OpenInherit.Sdk/Validator.cs
using System.Reflection;
using System.Text.Json;
using Json.Schema;

namespace OpenInherit.Sdk;

/// <summary>
/// Validates INHERIT documents against bundled JSON schemas (Draft 2020-12).
/// Schemas are embedded as resources so no file-system access is required at runtime.
/// </summary>
public static class Validator
{
    private const string CatalogueSchemaUri = "https://openinherit.org/v3/catalogue.json";

    /// <summary>
    /// Legal disclaimer that accompanies every validation result.
    /// </summary>
    public const string Disclaimer =
        "Validation results are informational. They verify schema " +
        "conformance and data structure, not legal accuracy or " +
        "completeness. Do not rely on validation results as the " +
        "sole basis for legal or financial decisions.";

    private static readonly Lazy<JsonSchema> EstateSchema =
        new(() => LoadEmbeddedSchema("inherit-v3-bundled.json"));

    private static readonly Lazy<JsonSchema> CatalogueSchema =
        new(() => LoadEmbeddedSchema("catalogue-v3-bundled.json"));

    /// <summary>
    /// Validates a JSON document, auto-detecting whether it is an estate or catalogue
    /// by inspecting its <c>$schema</c> property.
    /// </summary>
    public static ValidationResult Validate(string jsonDocument)
    {
        var mode = DetectMode(jsonDocument);
        return DoValidate(jsonDocument, mode);
    }

    /// <summary>
    /// Validates a JSON document as an INHERIT catalogue regardless of its <c>$schema</c> property.
    /// </summary>
    public static ValidationResult ValidateCatalogue(string jsonDocument)
        => DoValidate(jsonDocument, "catalogue");

    /// <summary>
    /// Validates a JSON document as an INHERIT estate regardless of its <c>$schema</c> property.
    /// </summary>
    public static ValidationResult ValidateEstate(string jsonDocument)
        => DoValidate(jsonDocument, "estate");

    /// <summary>
    /// Validates a JSON document at Level 2 (schema + referential integrity).
    /// First runs Level 1 (JSON Schema); if that fails the result has ConformanceLevel 0.
    /// If Level 1 passes, UUID cross-references between entities are checked.
    /// A document that passes both gets ConformanceLevel 2; one that passes Level 1 but has
    /// broken references stays at ConformanceLevel 1 with additional Level 2 errors.
    /// </summary>
    public static ValidationResult ValidateLevel2(string jsonDocument)
    {
        // Run Level 1 first.
        var mode = DetectMode(jsonDocument);
        var l1 = DoValidate(jsonDocument, mode);

        // If Level 1 failed, return immediately — no point checking refs.
        if (!l1.Valid)
            return l1;

        // Level 1 passed. Now check referential integrity.
        JsonDocument parsedDoc;
        try
        {
            parsedDoc = JsonDocument.Parse(jsonDocument);
        }
        catch
        {
            // Should be unreachable (Level 1 caught invalid JSON), but guard anyway.
            return l1;
        }

        using (parsedDoc)
        {
            var refErrors = ReferenceChecker.Check(parsedDoc.RootElement);

            if (refErrors.Count == 0)
            {
                return l1 with { ConformanceLevel = ConformanceLevel.Level2 };
            }

            // Level 2 failed: document is invalid, ConformanceLevel stays at 1.
            var allErrors = new List<ValidationError>(l1.Errors);
            allErrors.AddRange(refErrors);
            return l1 with { Valid = false, Errors = allErrors };
        }
    }

    // -------------------------------------------------------------------------

    private static string DetectMode(string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            if (doc.RootElement.TryGetProperty("$schema", out var schemaProp)
                && schemaProp.GetString() == CatalogueSchemaUri)
            {
                return "catalogue";
            }
        }
        catch
        {
            // Unparseable JSON — fall through; the validation step will surface the error.
        }
        return "estate";
    }

    private static ValidationResult DoValidate(string jsonDocument, string mode)
    {
        var schema = mode == "catalogue" ? CatalogueSchema.Value : EstateSchema.Value;

        JsonDocument parsedDoc;
        try
        {
            parsedDoc = JsonDocument.Parse(jsonDocument);
        }
        catch (JsonException ex)
        {
            return new ValidationResult(
                Valid: false,
                SchemaMode: mode,
                ConformanceLevel: ConformanceLevel.Failed,
                Errors: new[] { new ValidationError("/", $"Invalid JSON: {ex.Message}") },
                Disclaimer: Disclaimer
            );
        }

        using (parsedDoc)
        {
            var options = new EvaluationOptions
            {
                OutputFormat = OutputFormat.List
            };

            var result = schema.Evaluate(parsedDoc.RootElement, options);

            var errors = new List<ValidationError>();
            if (!result.IsValid)
            {
                // ToList() mutates result in-place, flattening the hierarchy so that
                // every failed leaf node appears in result.Details.
                result.ToList();

                foreach (var detail in result.Details ?? [])
                {
                    if (detail.IsValid || detail.Errors is null) continue;
                    var path = detail.InstanceLocation.ToString();
                    foreach (var kvp in detail.Errors)
                    {
                        errors.Add(new ValidationError(path, kvp.Value));
                    }
                }

                // Fallback: capture top-level errors if Details yielded nothing.
                if (errors.Count == 0 && result.Errors is not null)
                {
                    var path = result.InstanceLocation.ToString();
                    foreach (var kvp in result.Errors)
                    {
                        errors.Add(new ValidationError(path, kvp.Value));
                    }
                }

                // Last-resort sentinel so Errors is never empty on an invalid result.
                if (errors.Count == 0)
                {
                    errors.Add(new ValidationError("/", "Validation failed (no detailed error messages available)."));
                }
            }

            return new ValidationResult(
                Valid: result.IsValid,
                SchemaMode: mode,
                ConformanceLevel: result.IsValid ? ConformanceLevel.Level1 : ConformanceLevel.Failed,
                Errors: errors,
                Disclaimer: Disclaimer
            );
        }
    }

    private static JsonSchema LoadEmbeddedSchema(string resourceFileName)
    {
        var assembly = Assembly.GetExecutingAssembly();
        // MSBuild replaces hyphens with underscores in embedded resource names.
        var normalized = resourceFileName.Replace("-", "_");
        var resourceName = assembly
            .GetManifestResourceNames()
            .FirstOrDefault(n => n.EndsWith(resourceFileName, StringComparison.OrdinalIgnoreCase)
                              || n.EndsWith(normalized, StringComparison.OrdinalIgnoreCase))
            ?? throw new InvalidOperationException(
                $"Embedded resource '{resourceFileName}' not found. " +
                $"Available: {string.Join(", ", assembly.GetManifestResourceNames())}");

        using var stream = assembly.GetManifestResourceStream(resourceName)!;
        using var reader = new StreamReader(stream);
        var json = reader.ReadToEnd();

        // Use a fresh local SchemaRegistry so that sub-schema $id URIs from this
        // bundled file do not collide with those from the other bundled file in the
        // global registry when both schemas are loaded in the same process.
        var buildOptions = new BuildOptions
        {
            SchemaRegistry = new SchemaRegistry()
        };
        return JsonSchema.FromText(json, buildOptions);
    }
}
